visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.service.ts

```diff
@@ -0,0 +1,111 @@
+import { StoreService, StoreSubscriber } from "../../../store.service"
+import { IRootScopeService } from "angular"
+import { calculateNewEdgeMetricData, EdgeMetricDataActions } from "./edgeMetricData.actions"
+import { isActionOfType } from "../../../../util/reduxHelper"
+import {
+	AttributeTypes,
+	AttributeTypeValue,
+	BlacklistItem,
+	CodeMapNode,
+	EdgeMetricCount,
+	EdgeMetricData
+} from "../../../../codeCharta.model"
+import { FileState } from "../../../../model/files/files"
+import { HierarchyNode } from "d3"
+import { BlacklistService, BlacklistSubscriber } from "../../fileSettings/blacklist/blacklist.service"
+import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
+import { nodeEdgeMetricsMap } from "./edgeMetricData.reducer"
+import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
+
+export interface EdgeMetricDataSubscriber {
+	onEdgeMetricDataChanged(edgeMetricData: EdgeMetricData[])
+}
+
+export class EdgeMetricDataService implements StoreSubscriber, BlacklistSubscriber, FilesSelectionSubscriber, AttributeTypesSubscriber {
+	private static EDGE_METRIC_DATA_CHANGED_EVENT = "edge-metric-data-changed"
+	public static NONE_METRIC = "None"
+
+	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+		StoreService.subscribe(this.$rootScope, this)
+		BlacklistService.subscribe(this.$rootScope, this)
+		FilesService.subscribe(this.$rootScope, this)
+		AttributeTypesService.subscribe(this.$rootScope, this)
+	}
+
+	public onStoreChanged(actionType: string) {
+		if (isActionOfType(actionType, EdgeMetricDataActions)) {
+			this.notify(this.select())
+		}
+	}
+
+	public onBlacklistChanged(blacklist: BlacklistItem[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(this.storeService.getState().files, blacklist))
+	}
```

Comment:
These store dispatches happen very frequently. I personally would expect that it's possible to abstract these somehow to have a simple function that does the dispatching for different kinds of metrics. But I did not check if that is really useful here.

Reply:
Of course store dispatches happen a lot. That's the redux way of changing the state and notifying all subscribers to update their viewModel.

Reply:
What I meant was the dispatch declaration happens frequently with very similar patterns and we might be able to consolidate some of those to use a single function for dispatching multiple to reduce the code overhead and simplify the implementation.

Reply:
The only case where the dispatch happens frequently is, when we load a new map. In this case, we first trigger `onFilesSelectionChanged`. After that, `onAttributeTypesChanged` and `onBlacklistChanged` are triggered, because we have to reset the blacklist and update the attributeTypes with those from the new map.

In that case, combining them to a single dispatch would be great, but I don't really know how we can do this. We could use a `debounce-function` to wait 200ms for more events before dispatching.